quic/s2n-quic-transport/src/ack/ack_ranges.rs

```diff
@@ -0,0 +1,241 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::interval_set::{IntervalSet, RangeInclusiveIter};
+use core::{
+    convert::TryInto,
+    num::NonZeroUsize,
+    ops::{Bound, Deref, DerefMut, RangeInclusive},
+};
+use s2n_quic_core::{
+    ack::Settings,
+    frame::ack,
+    packet::number::{PacketNumber, PacketNumberRange},
+    varint::VarInt,
+};
+
+#[derive(Clone, Debug)]
+pub struct AckRanges(IntervalSet<PacketNumber>);
+
+impl Default for AckRanges {
+    fn default() -> Self {
+        Self::new(Settings::default().ack_ranges_limit as usize)
+    }
+}
+
+impl AckRanges {
+    pub fn new(limit: usize) -> Self {
+        let limit = NonZeroUsize::new(limit).expect("limit should be nonzero");
+        Self(IntervalSet::with_limit(limit))
+    }
+
+    /// Inserts a packet number; dropping smaller values if needed
+    #[inline]
+    pub fn insert_packet_number_range(&mut self, pn_range: PacketNumberRange) -> Result<(), ()> {
+        let interval = (
+            Bound::Included(pn_range.start()),
+            Bound::Included(pn_range.end()),
+        );
+        if self.0.insert(interval).is_ok() {
+            return Ok(());
+        }
+
+        // attempt to shed the lowest packet number ranges to make room for larger ones
+        match self.0.pop_min() {
+            Some(min) => {
+                if min < pn_range.start() {
+                    // TODO: add metrics for ack ranges being dropped
+                    self.0.insert(interval).expect(
+                        "min range was removed, so it should be possible to insert another range",
+                    );
+                    Ok(())
```

Comment:
Can I get another pair or eyes verifying that this is correct.

Previously this was doing `self.0.insert(interval).is_ok()`